cli_tools/common/utils/logging/logger.go

```diff
@@ -15,50 +15,205 @@
 package logging
 
 import (
-	"fmt"
 	"log"
+	"os"
+	"strings"
+	"sync"
 	"time"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/proto/go/pb"
+	"github.com/golang/protobuf/proto"
 )
 
-// LogEntry encapsulates a single log entry.
-type LogEntry struct {
-	LocalTimestamp time.Time `json:"localTimestamp"`
-	Message        string    `json:"message"`
+// LogWriter is a logger for interactive tools. It supports
+// string messages and structured metrics.
+//
+// Structured metrics are accumulated over the lifespan of the logger.
+//
+// To rebuild the mock, run `go generate ./...`
+//go:generate go run github.com/golang/mock/mockgen -package mocks -source $GOFILE -destination ../../../mocks/mock_logger.go
+type LogWriter interface {
+	// WriteUser messages appear in the following places:
+	//  1. Web UI and gcloud.
+	//  2. Standard output of the CLI tool.
+	//  3. Backend trace logs (all debug and user logs are combined to a single trace log).
+	WriteUser(message string)
+	// WriteDebug messages appear in the following places:
+	//  1. Standard output of the CLI tool.
+	//  2. Backend trace logs (all debug and user logs are combined to a single trace log).
+	WriteDebug(message string)
+	// WriteTrace messages are saved to the logging backend (OutputInfo.serial_outputs).
+	WriteTrace(message string)
+	// WriteMetric merges all non-default fields into a single OutputInfo instance.
+	WriteMetric(metric *pb.OutputInfo)
+}
+
+// RedirectGlobalLogsToUser redirects the standard library's static logger
+// to logWriter.User.
+func RedirectGlobalLogsToUser(logWriter LogWriter) {
+	log.SetPrefix("")
+	log.SetFlags(0)
+	log.SetOutput(redirectShim{logWriter})
+}
+
+// redirectShim forwards Go's standard logger to LogWriter.User.
+type redirectShim struct {
+	writer LogWriter
+}
+
+func (l redirectShim) Write(p []byte) (n int, err error) {
+	l.writer.WriteUser(string(p))
+	return len(p), nil
+}
+
+// LogReader exposes pb.OutputInfo to a consumer.
+type LogReader interface {
+	ReadOutputInfo() *pb.OutputInfo
+}
+
+// ToolLogger is a logger for interactive tools. It supports
+// string messages and structured metrics.
+//
+// Structured metrics are accumulated over the lifespan of the logger.
+type ToolLogger interface {
```

Comment:
If this was named just `Logger` and had all functions from the two interfaces inlined, it would be a simpler structure.Related to the question below on the split between these two interfaces.

Reply:
Resolving since the thread "I don't see an obvious benefits to split the interface to writer+reader." is open for this s